### PR TITLE
Vote history 1

### PR DESCRIPTION
**Pull request by NA-CA**

### CODE REVIEW

File: src/gameplay/game.ts [hunk 1]
```diff
@@ -990,7 +987,16 @@ class Game extends Room {
...
-        data[i].voteHistory = this.voteHistory;
+        if (this.closeVoteHistory)
+        {
+          data[i].voteHistory = {}; 
+        }
+        else
+        {
+          data[i].voteHistory = this.voteHistory;
+
+        }
```
Comment (vck3000) [inline src/gameplay/game.ts:998]: You can change all of your code above into just this:

```
data[i].voteHistory = this.closeVoteHistory ? {} : this.voteHistory
```

Comment (vck3000) [reply src/gameplay/game.ts:998]: This is the ternary operator btw. Super handy in simple cases like these!

File: src/gameplay/game.ts [hunk 2]
```diff
@@ -1860,11 +1869,22 @@ class Game extends Room {
...
 
+  updateCloseVoteHistory(closeVoteHistory: boolean) {
+    this.closeVoteHistory = closeVoteHistory;
+
+    this.sendText(
+      this.allSockets,
+      `Close vote history option set to ${closeVoteHistory}.`,
+      'server-text',
+    );
+
```
Comment (vck3000) [inline src/gameplay/game.ts:1880]: Nit: can we delete this space? Have you set up prettier and auto-format on save?

File: src/gameplay/game.ts [hunk 3]
```diff
@@ -1949,17 +1969,14 @@ class Game extends Room {
...
   If there is only one provisional player in the game:
   Formula: R_new = (R_old*N_old + sum(otherPlayerRatings)/numOtherPlayers + 200*TeamAdjustment*Result)/(N_old+1)
       where N_old = Number of games played
             Result = 1 for win, 0 for loss
-
   If there is more than one provisional player in the game:
   Formula: R_new = (R_old*N_old + sum(allPlayerRatings)/numPlayers + 200*TeamAdjustment*Result)/(N_old+1)
       where N_old = Number of games played
             Result = 1 for win, 0 for loss
-
```
Comment (vck3000) [inline src/gameplay/game.ts:1962]: Let's add spaces back in pls!

File: src/modsadmins/mods.ts [hunk 4]
```diff
@@ -5,7 +5,6 @@ export const modsArray = [
   'not.me',
   'manuavalon',
   'maximovic96',
-  'd8nitemike',
```
Comment (vck3000) [inline src/modsadmins/mods.ts:8]: Uhh, not sure why this change is here. Can you pull master to make sure it's up to date in your branch?

File: package-lock.json
Comment (vck3000) [inline package-lock.json:1]: Oh, this file isn't git-ignored. Could you add it please? We should only be using yarn.lock :)

**Review by vck3000 — CHANGES_REQUESTED** [review]

File: assets/scripts/lobby/buttonEventListeners2.js [hunk 5]
```diff
@@ -185,13 +185,15 @@ $('#createNewRoomButton').on('click', (data) => {
...
     muteSpectators: $('.muteSpectators')[1].checked,
+    closeVoteHistory: $('.closeVoteHistory')[1].checked,
     ranked: $($('.rankedSelect')[1]).val(),
   };
 
   // Update the settings in the in room settings menu.
   $($('.maxNumPlayers')[0]).val(sendObj.maxNumPlayers);
   $($('.gameModeSelect')[0]).val(sendObj.gameMode);
   $('.muteSpectators')[0].checked = sendObj.muteSpectators;
+  $('.closeVoteHistory')[0].checked = sendObj.closeVoteHistory;
```
[the remark this answers is not in the document]
Comment (vck3000) [reply assets/scripts/lobby/buttonEventListeners2.js:196]: You can do this easily with a find replace across all files :)

File: assets/scripts/lobby.js [hunk 6]
```diff
@@ -1322,6 +1328,13 @@ function generatePlayerLocations(numOfPlayers, a, b) {
 
 // Note this function will also draw the card history
 function drawVoteHistory(data) {
+  console.log(data);
+  console.log(data.voteHistory, 'tiger');
+
+  if ((data.voteHistory = {})) {
```
Comment (vck3000) [inline assets/scripts/lobby.js:1334]: Here's your issue!!! You're assigning, not comparing!

File: package-lock.json
Comment (vck3000) [inline package-lock.json:1]: Please remove this file and add it to the `.gitignore`!

File: assets/scripts/lobby.js [hunk 7]
```diff
@@ -86,6 +86,8 @@ let highlightedAvatars;
 function draw() {
   // console.log("draw called");
   if (roomPlayersData) {
+
+    console.log(roomPlayersData, 'rpd');
```
Comment (vck3000) [inline assets/scripts/lobby.js:90]: Remove please :)

File: assets/scripts/lobby.js [hunk 8]
```diff
@@ -1322,6 +1328,13 @@ function generatePlayerLocations(numOfPlayers, a, b) {
 
 // Note this function will also draw the card history
 function drawVoteHistory(data) {
+  console.log(data);
+  console.log(data.voteHistory, 'tiger');
```
Comment (vck3000) [inline assets/scripts/lobby.js:1332]: Remove please :)

File: assets/scripts/lobby.js [hunk 9]
```diff
@@ -1360,6 +1373,16 @@ function drawVoteHistory(data) {
...
     // for every mission
+
+    //console.log(data.voteHistory[keyArray[k]].length);
+
+    /*
+    if (data.voteHistory[keyArray[k]].length === 0)
+    {
+      return; 
+    }
+    */
```
Comment (vck3000) [inline assets/scripts/lobby.js:1384]: Remove please :)

File: assets/scripts/lobby/sockets/game.js [hunk 10]
```diff
@@ -239,9 +239,19 @@ function removeTeamHighlightAndLeaderOutline() {
 // Given a mission number and the pick number, returns the people on that mission pick
 // If pickNum is -1, then returns the people on the last pick of that mission
 function getPlayersOnMissionPickAndLeader(missionNum, pickNum = -1) {
+  if ((gameData.voteHistory = {})) {
```
Comment (vck3000) [inline assets/scripts/lobby/sockets/game.js:242]: Please remove the extra brackets

File: assets/scripts/lobby/sockets/game.js [hunk 11]
```diff
@@ -239,9 +239,19 @@ function removeTeamHighlightAndLeaderOutline() {
...
+  }
   // We need a player key to see vote history to get the number of picks in the mission
   const firstPlayerKey = Object.keys(gameData.voteHistory)[0];
 
+  /*
+  if(gameData.voteHistory[firstPlayerKey].length === 0)
+  {
+    return; 
+  }
+  */
```
Comment (vck3000) [inline assets/scripts/lobby/sockets/game.js:253]: Remove please :)

File: assets/scripts/lobby/sockets/sockets.js [hunk 12]
```diff
@@ -271,7 +271,8 @@ socket.on('update-current-players-list', (currentPlayers) => {
   rankedPlayers.forEach((player) => {
     // if the current player exists, add it
     if (player) {
-      str = `<tr> <td>${player.displayUsername} ${generateBadgeString(player.badge)}</td> <td align="right">${player.ratingBadge} ${player.playerRating}</td> </tr>`;
+      str = `<tr> <td>${player.displayUsername} ${generateBadgeString(player.badge,)}</td> <td align="right">${player.ratingBadge} 
+      ${player.playerRating}</td> </tr>`;
```
Comment (vck3000) [inline assets/scripts/lobby/sockets/sockets.js:275]: I don't think this is intentional. It could mess up the UI. Could you revert this change please?

File: assets/scripts/lobby.js [hunk 13]
```diff
@@ -994,7 +1000,7 @@ function strOfAvatar(playerData, alliance) {
       ).width + 20;
 
     role = `<p class='role-p' style='width: ${roleWid}px; margin: auto;'>${gameData.see.roles[getIndexFromUsername(playerData.username)]
-      }</p>`;
+    }</p>`;
```
Comment (vck3000) [inline assets/scripts/lobby.js:1003]: There's still a few formatting changes that need to be reverted here

File: src/gameplay/game.ts [hunk 14]
```diff
@@ -990,7 +995,13 @@ class Game extends Room {
         data[i].numSelectTargets = this.getClientNumOfTargets(i);
 
         data[i].votes = this.publicVotes;
-        data[i].voteHistory = this.voteHistory;
+
+        console.log(this.disableVoteHistory, 'cvh');
```
Comment (vck3000) [inline src/gameplay/game.ts:999]: Remove please :)

File: src/gameplay/game.ts [hunk 15]
```diff
@@ -1851,6 +1862,9 @@ class Game extends Room {
   }
 
   updateMuteSpectators(muteSpectators: boolean) {
+    console.log(muteSpectators);
+    console.log('hello');
```
Comment (vck3000) [inline src/gameplay/game.ts:1866]: Remove please :)

File: src/gameplay/game.ts [hunk 16]
```diff
@@ -1860,11 +1874,21 @@ class Game extends Room {
     );
   }
 
+  updatedisableVoteHistory(disableVoteHistory: boolean) {
+    this.disableVoteHistory = disableVoteHistory;
+
+    console.log(disableVoteHistory, 'test');
```
Comment (vck3000) [inline src/gameplay/game.ts:1880]: Remove please :)

File: src/models/gameRecord.js [hunk 17]
```diff
@@ -35,6 +35,7 @@ const gameRecordSchema = new mongoose.Schema({
   missionHistory: [String],
   numFailsHistory: [Number],
   voteHistory: Object,
+  disableVoteHistory: Boolean,
```
Comment (vck3000) [inline src/models/gameRecord.js:38]: You'll also need to add to this line here in `finishGame()` to include the data:

https://github.com/vck3000/ProAvalon/blob/master/src/gameplay/game.ts#L1278-L1316

**Review by vck3000 — CHANGES_REQUESTED** [review]

File: src/models/gameRecord.js (hunk 17 above)
Comment (NA-CA) [reply src/models/gameRecord.js:38]: Fixed this.

File: src/modsadmins/mods.ts (hunk 4 above)
Comment (NA-CA) [reply src/modsadmins/mods.ts:8]: Fixed this as well.

File: package-lock.json
Comment (vck3000) [inline package-lock.json:1]: Please remove this and add to gitignore

File: assets/scripts/lobby/sockets/game.js [hunk 18]
```diff
@@ -239,6 +238,9 @@ function removeTeamHighlightAndLeaderOutline() {
 // Given a mission number and the pick number, returns the people on that mission pick
 // If pickNum is -1, then returns the people on the last pick of that mission
 function getPlayersOnMissionPickAndLeader(missionNum, pickNum = -1) {
+  if (gameData.voteHistory = {}) {
```
Comment (vck3000) [inline assets/scripts/lobby/sockets/game.js:241]: This is wrong?

File: src/gameplay/game.ts [hunk 19]
```diff
@@ -1860,6 +1873,22 @@ class Game extends Room {
...
+      this.disableVoteHistory = disableVoteHistory;
+
+      this.sendText(
+        this.allSockets,
+        `Close Vote History option set to ${disableVoteHistory}.`,
+        'server-text',
+      );
+
+    }
+
```
Comment (vck3000) [inline src/gameplay/game.ts:1889]: Can remove the extra new lines here

File: src/modsadmins/mods.ts (hunk 4 above)
Comment (vck3000) [reply src/modsadmins/mods.ts:8]: It's still showing differences here. You sure you pulled in master?

File: src/gameplay/game.ts [hunk 20]
```diff
@@ -1860,6 +1873,22 @@ class Game extends Room {
     );
   }
 
+  updatedisableVoteHistory(disableVoteHistory: boolean) {
```
Comment (vck3000) [inline src/gameplay/game.ts:1876]: updateDisableVoteHistory

File: yarn.lock
Comment (vck3000) [inline yarn.lock:1]: This shouldn't change either. Can you revert this file? This only changed earlier because you added a package earlier which you've now removed.

File: src/models/gameRecord.js (hunk 17 above)
Comment (vck3000) [reply src/models/gameRecord.js:38]: I don't see this added yet?

**Review by vck3000 — CHANGES_REQUESTED** [review]

File: src/models/gameRecord.js (hunk 17 above)
Comment (NA-CA) [reply src/models/gameRecord.js:38]: https://github.com/vck3000/ProAvalon/pull/509/files#diff-fa564d9f0a3077d128a9edae956714e418036440f018e4c42849d44fca697ab4R1315

File: src/modsadmins/mods.ts (hunk 4 above)
Comment (NA-CA) [reply src/modsadmins/mods.ts:8]: Yep, this file matches the existing file in the code, I checked it.

File: assets/scripts/lobby.js [hunk 21]
```diff
@@ -1322,6 +1327,11 @@ function generatePlayerLocations(numOfPlayers, a, b) {
 
 // Note this function will also draw the card history
 function drawVoteHistory(data) {
+
+  if ((data.voteHistory === {})) {
```
[the remark this answers is not in the document]
Comment (NA-CA) [reply assets/scripts/lobby.js:1326]: Yea we can, I changed it.

File: assets/scripts/lobby/sockets/game.js (hunk 18 above)
Comment (NA-CA) [reply assets/scripts/lobby/sockets/game.js:241]: fixed mb

File: package-lock.json
Comment (NA-CA) [reply package-lock.json:1]: ok will do.

File: src/gameplay/game.ts (hunk 20 above)
Comment (NA-CA) [reply src/gameplay/game.ts:1876]: fixed.

File: src/gameplay/game.ts (hunk 19 above)
Comment (NA-CA) [reply src/gameplay/game.ts:1889]: fixed.

File: yarn.lock
Comment (NA-CA) [reply yarn.lock:1]: followed instructions on discord.

File: src/modsadmins/mods.ts (hunk 4 above)
Comment (vck3000) [reply src/modsadmins/mods.ts:8]: Ok all good

**Comment by vck3000** [comment]
It's look good to me now! I will give it a local test and make sure everything works. 

Last thing, would you like to add a changelog entry? :). It'll be in `src/views/log.ejs`

**Review by vck3000 — APPROVED** [review]

File: src/gameplay/game.ts [hunk 22]
```diff
@@ -990,7 +1000,9 @@ class Game extends Room {
         data[i].numSelectTargets = this.getClientNumOfTargets(i);
 
         data[i].votes = this.publicVotes;
-        data[i].voteHistory = this.voteHistory;
+
+        data[i].voteHistory = this.disableVoteHistory ? {} : this.voteHistory;
```
Comment (vck3000) [inline src/gameplay/game.ts:1004]: This should be `? null : this.voteHistory` I think!
